### PR TITLE
Issue #6345: RightCurly with option alone false negative for class, method and constructor

### LINKED ISSUE #6345: RightCurly with option alone false negative for class, method and constructor
```
/var/tmp $ javac MyClass.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="RightCurly">
            <property name="option" value="alone"/>
            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF"/>
        </module>
    </module>
</module>

/var/tmp $ cat MyClass.java
public final class MyClass {
    public MyClass() {} // Doesn't warn - incorrect
    
    public void main() {} // Doesn't warn - incorrect
    
    class Foo {} // Doesn't warn - incorrect
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-8.16-all.jar -c config.xml MyClass.java
Starting audit...
Audit done.
```

---------------

In all three cases (constructor, method, class), the right curly brace is on the same line as the left curly brace. However, the option `alone` should force the right curly brace to be on it's own line,

Correct:
```java
public final class MyClass {
    public MyClass() {
    }
    
    public void main() {
    }
    
    class Foo {
    }
}
```

---------------
https://groups.google.com/forum/#!topic/checkstyle/q_kVS42ITt0

### PR DESCRIPTION
**Pull request by fzdy1914**
Fixes #6345.

When I working with this pr, I found that this may be an intended behavior of checkstyle.
See [here](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L225), it allows empty body.

If it is not the intended behavior, I will just delete the line that allows empty body.

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 1]
```diff
@@ -221,8 +221,7 @@ private static boolean shouldBeAloneOnLine(RightCurlyOption bracePolicy,
      */
     private static boolean shouldBeAloneOnLineWithAloneOption(Details details,
                                                               String targetSrcLine) {
-        return !isAloneOnLine(details, targetSrcLine)
-                && !isEmptyBody(details.lcurly);
+        return !isAloneOnLine(details, targetSrcLine);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: I have tracked this `isEmptyBody` line back to https://github.com/checkstyle/checkstyle/commit/f1dced161ab0c818f0e1e125fc5b88b71f5b171e#diff-9b2a5f9e69b254b923cb4fd7420cb9a4R187
````
         else if (getAbstractOption() == RightCurlyOption.ALONE
-                && rcurly.getLineNo() == nextToken.getLineNo()
+                && !isAloneOnLine(details)
+                && !isEmptyBody(lcurly)) {
````

@romani It seems this was specifically requested and allowed in https://github.com/checkstyle/checkstyle/issues/1019 .
Please review.

Comment (fzdy1914) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: However, I believe this line is more like a mistake. The goal of that pr is to add a new option `ALONE_OR_SINGLELINE`, but it modified the behavior of `ALONE`. So maybe it is indeed a mistake.

Comment (fzdy1914) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: This [comment](https://github.com/checkstyle/checkstyle/blame/5a65f636aa6b08321ac8900bb1214a1a0fd410d6/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/rightcurly/InputRightCurlyLeft.java#L144) is weird also.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Your are correct @fzdy1914 . Looking at the issue more closely, it is just introducing that option and wasn't suppose to be changing `alone` from what I can see. Changes were noticed in the PR at https://github.com/checkstyle/checkstyle/pull/1259#issuecomment-116502766 but were deemed good and not fully related to this issue.

I don't see any reason why we shouldn't continue this PR.
@romani if you are ok with everything, mark this discussion as resolved.

**Comment by rnveach** [comment]
@fzdy1914 Please make all CI green and keep number of commits at 1.

**Comment by fzdy1914** [comment]
@rnveach Can you help look into the failing test?

**Comment by rnveach** [comment]
@romani We may have to help fix wercker failure since this PR may affect other projects.

**Comment by rnveach** [comment]
@fzdy1914 
pitest failure is:
> `RightCurlyCheck.java.html:<td class='covered'><pre><span  class='survived'>            if (lcurly.getNextSibling().getType() == TokenTypes.RCURLY) {</span></pre></td></tr>`

You will have to run `./.ci/pitest.sh pitest-blocks` from base checkstyle directory to be able to see report and find where this line is.
If your not familar with pitest, which is mutation test, it is basically modifying your code to look for tests failure. It found that this line was modified and no tests were failing. Something was failing before, so your change some how led to this.

Let me know if you need more help.

**Comment by fzdy1914** [comment]
Ready for review.

File: .ci/pitest.sh [hunk 2]
```diff
@@ -227,6 +227,7 @@ pitest-blocks)
   "RightCurlyCheck.java.html:<td class='covered'><pre><span  class='survived'>                &#38;&#38; details.nextToken.getType() == TokenTypes.RCURLY</span></pre></td></tr>"
   "RightCurlyCheck.java.html:<td class='covered'><pre><span  class='survived'>                &#38;&#38; tokenAfterNextToken.getType() == TokenTypes.SEMI</span></pre></td></tr>"
   "RightCurlyCheck.java.html:<td class='covered'><pre><span  class='survived'>            else if (tokenType == TokenTypes.LITERAL_CATCH) {</span></pre></td></tr>"
+  "RightCurlyCheck.java.html:<td class='covered'><pre><span  class='survived'>            if (lcurly.getNextSibling().getType() == TokenTypes.RCURLY) {</span></pre></td></tr>"
```
Comment (rnveach) [inline .ci/pitest.sh]: Why can't you try and cover this line?
We are trying to kill all mutants.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 3]
```diff
@@ -235,8 +234,8 @@ private static boolean shouldBeAloneOnLineWithAloneOption(Details details,
     private static boolean shouldBeAloneOnLineWithAloneOrSinglelineOption(Details details,
                                                                           String targetSrcLine) {
         return !isAloneOnLine(details, targetSrcLine)
-                && !isSingleLineBlock(details)
-                && !isEmptyBody(details.lcurly);
+                && !isEmptyBody(details.lcurly)
+                && !isSingleLineBlock(details);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Please explain this change.

Comment (fzdy1914) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: This is for the coverage. Previously, `isEmptyBody` is also used in `shouldBeAloneOnLineWithAloneOption`, so the coverage is guaranteed. 

Now, it is the only place for this method, so the order is important.

To cover one of line in  `isEmptyBody`, the following line is added.
`class TestClass4 { }`
However, `!isSingleLineBlock(details)` will return false in the first place,  so this line cannot be passed in the `isEmptyBody` function anymore.

Let's reorder them so that such line will be examined in the first place.

File: .ci/pitest.sh (hunk 2 above)
Comment (fzdy1914) [reply .ci/pitest.sh]: OK, I will try to find if there is a test to cover this.

**Comment by fzdy1914** [comment]
Ready to review.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 3 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: This is no longer re-ordered so I am assume it was resolved so I will mark this as resolved.

**Comment by rnveach** [comment]
@romani please review and review https://github.com/checkstyle/checkstyle/pull/6505#discussion_r262122858 .

**Comment by fzdy1914** [comment]
@romani Can kindly give your review for this PR?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: it was mistake to affect `alone` mode

**Comment by rnveach** [comment]
@fzdy1914 as PR is entering final stages, please provide us a regression report. See https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#diffgroovy-diff-report-generation .
Please review the report and ensure you see no issues being reported and we will review it too.

**Comment by romani** [comment]
I am ok with code changed.
@fzdy1914 , thanks a lot, and sorry for long delay in reply.
Regression report is required to make sure we did not do extra damage. Please do regression report for one modes of Check to make sure we did not damage another mode (as it was in original issue).

**Comment by fzdy1914** [comment]
> I recommend creating a configuration with as many possible property mutations to ensure we don't miss anything. You can assign each instance of a config with an id. See https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-pitest-regression with an example config.


What does property mutation mean? Do you mean that I should write a config to test all possible combinations of tokens and mode to test if there is a regression?

**Comment by fzdy1914** [comment]
I wrote the following checks and get an error, can you help to examine my file.
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">


<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
         <!-- Example of sevntu.checkstyle Check usage -->
         <!-- <module name="NestedSwitchCheck"/> -->

         <!-- Example of checkstyle Check usage -->
         <module name="ThrowsCount">
             <property name="max" value="20000000"/>
         </module>

        <module name="RightCurly">
            <property name="option" value="alone"/>
            <property name="tokens" value="LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY,
            LITERAL_IF, LITERAL_ELSE, CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE,
            LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
            <property name="id" value="RCurly Alone All Tokens" />
        </module>

        <module name="RightCurly">
            <property name="option" value="alone"/>
            <property name="tokens" value="LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE"/>
            <property name="id" value="RCurly Alone Part Tokens 1" />
        </module>

         <module name="RightCurly">
            <property name="option" value="alone"/>
            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE,
            LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
            <property name="id" value="RCurly Alone Part Tokens 2" />
        </module>

        <module name="RightCurly">
            <property name="option" value="same"/>
            <property name="tokens" value="LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY,
            LITERAL_IF, LITERAL_ELSE, CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE,
            LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
            <property name="id" value="RCurly Same All Tokens" />
        </module>

        <module name="RightCurly">
            <property name="option" value="same"/>
            <property name="tokens" value="LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE"/>
            <property name="id" value="RCurly Same Part Tokens 1" />
        </module>

         <module name="RightCurly">
            <property name="option" value="same"/>
            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE,
            LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
            <property name="id" value="RCurly Same Part Tokens 2" />
        </module>

        <module name="RightCurly">
            <property name="option" value="alone_or_singleline"/>
            <property name="tokens" value="LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY,
            LITERAL_IF, LITERAL_ELSE, CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE,
            LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
            <property name="id" value="RCurly Alone or SingleLine All Tokens" />
        </module>

        <module name="RightCurly">
            <property name="option" value="alone_or_singleline"/>
            <property name="tokens" value="LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE"/>
            <property name="id" value="RCurly Alone or SingleLine Part Tokens 1" />
        </module>

         <module name="RightCurly">
            <property name="option" value="alone_or_singleline"/>
            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE,
            LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
            <property name="id" value="RCurly Alone or SingleLine Part Tokens 2" />
        </module>
    </module>
<!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
-->
</module>
```
image

**Comment by rnveach** [comment]
> Do you mean that I should write a config to test all possible combinations of tokens and mode to test if there is a regression?

Combinations of properties. It is ok if after the default check you specify all tokens.

> can you help to examine my file.

> `<module name="ThrowsCount">`

You can remove this as it isn't needed. It was just an example.

> Property `cacheFile` does not exist

There is no problem with your config since you don't specify this. It looks to be an issue with maven-site (or something on the maven side).
I will have to look into this more.

**Comment by rnveach** [comment]
@fzdy1914 Please check that you are using the latest master in contribution. Your output is saying maven-checkstyle-plugin 2.17 and we are using 3.0 . This is the cause of the failure.

Also,  as another recommendation, the more modules you put into a config the more memory you will have to give regression. To ensure you don't run into memory issues and have to restarted regression I usually give it 3 gigs of ram or split up the regression into 3 for each option.

**Comment by romani** [comment]
It is better to have only one Check in config for one report generation, it will be easier to verify it. You can run diff tool few times, have few reports.

**Comment by romani** [comment]
I reproduced the same problem with cacheFile, fix is merged to contribution repo, please pull from us code to avoid problem.

**Comment by fzdy1914** [comment]
This is the report generated: [link](https://fzdy1914.github.io/6345/index.html). It seems ok from my observation.

**Comment by romani** [comment]
@fzdy1914 , there are CI failures in tests, please fix and relaunch regression

**Comment by fzdy1914** [comment]
Okay, seems my other PR causes some problem with this one.

**Comment by fzdy1914** [comment]
@romani There is no need to regenerate the report as I only need to add [one line](https://github.com/checkstyle/checkstyle/blob/56454135ed5e8c2a1dd184108257a2ec662e96cb/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheckTest.java#L455) of 'test code' to fix this.
The `main code` is not touched.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/DetectorOptions.java [hunk 4]
```diff
@@ -54,7 +54,8 @@
     private Pattern pattern;
 
     /** Default constructor.*/
-    private DetectorOptions() { }
+    private DetectorOptions() {
+    }
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/DetectorOptions.java]: please move this change to separate commit (first commit) of this PR.
Such change is ok even without fix in Check. It is ok to keep reference to the same issue as a reason of update.

**Comment by fzdy1914** [comment]
Changed as request
